### ISSUE TITLE
dropped nose-tip to nose-tip frames on mice home cage

### ISSUE THREAD

**Issue by donw7** [murthylab/sleap#423]
Hello, thanks for the great work! Tracking is working pretty well for us except with one consistent issue - in frames where mice are interacting nose-tip to nose-tip, it seems to lose one nose in about 30% of these frames. The rest of the skeleton is tracked fine. This happens even when there does not appear to be occlusion; and there are numerous training examples of this type of interaction. See below for example image. Any help on this would be much appreciated. 

(used approx 3000 examples, with about 1-2000 during close interaction; skeleton designed outward from center 'scopebase' w/ nodes {nose, scopebase, scopelight, earL, earR, tail}; bottom-up, sigma for nodes 7, sigma for edges 50, input scaling 2, batch size 5,  learning rate 0.0005, with all augmentation turned on)

![image](https://user-images.githubusercontent.com/36974655/100821321-fcc1b980-3404-11eb-93cc-6065d233d207.png)

**Comment by talmo** [comment]
Hi @donw7,

Thanks for the detailed information!

So let's see. Perhaps you can try a couple of things:

1. Try increasing the sigma for edges to 75, and possibly check out what they look like by clicking on *Predict* -> *Run training...* -> *View Training Image Inputs...*. At 1280 x 1024 at ~1.9 pixels/mm, this is what PAFs with sigma = 50 looks like:
![image](https://user-images.githubusercontent.com/3187454/100827450-bc753200-342a-11eb-84cb-822ec8731de9.png)

You might have a bit better chance of not losing the nose tip if you have a little more support around that area.

The other thing you can do is decrease the sigma for the nodes slightly to 4-5. This will results in more precise localization of the nose points and consequently reduce the chance that they'll be as closely overlapping.


2. You should definitely not have the input scaling > 1.0. This will resize your images to be bigger, but it doesn't add any information and increases the memory requirements for the model. I'd leave it at 1.0 or possibly even 0.75.


3. You can try a top-down model for reference. Some of our top-down mouse models performed nearly as well as bottom-up, and it may be that in your data it won't suffer from the nose-tip issue that you're getting now. Be sure to choose an appropriate anchor node to center on (perhaps the 'scopebase').


There's more things to try but let's start with that. Let me know what you get and we'll go from there.

Cheers :)

**Comment by donw7** [comment]
Ok great thanks so much for the detailed and prompt response! Will try it and let you know. It sounds like you're saying increasing sigma for edges allows more PAF structure to connect to the nose area, while decreasing sigma for nodes increases precision - however, the visualization seems to show the confidence maps as becoming smaller (more nonoverlapping) with increased sigma to 75? 

One thing I forgot to mention is I believe we have a mix of labeling nonvisible or occluded body parts as nonvisible explicitly (w/ right click) or simply keeping the same label with a guess about where it might be. They are mostly labeled as default visible (the latter), but overall the training examples are not strictly consistent w/ toggling nonvisibility - would this matter meaningfully for this issue?

**Comment by talmo** [comment]
Hi @donw7,

Sorry for the delay!

So the sigma for the PAFs should not affect the size of the confidence maps, but yes, I'm proposing changing the two in opposite directions -- the PAFs bigger and the confidence maps smaller.

Regarding the labeling consistency: That can definitely introduce problems. Being inconsistent about labeling a part as visible/not visible can sometimes introduce really weird behavior because the neural network will try to figure out what's special about the images in each case, and if it turns out there's nothing then it'll end up focusing on small, irrelevant details that correlate with the labels being visible/not visible.

I would definitely highly recommend trying to do a pass through your labels to ensure they are consistent with respect to the visibility of the body parts. There's a good possibility it plays a role into this issue (though it may not be the only factor), so if you can eliminate that possibility we'll be better off with troubleshooting :)

**Comment by donw7** [comment]
Np. Thanks so much for your help. 

I corrected the inconsistency issue. It seems changing the sigmas for bottom up helped a bit but still quite a few frames missing one nose tip. Top down seems to work okay in terms of not dropping nose tips and accuracy overall, however from frame to frame there appears to be a jitter (i.e. the nodes shift a few pixels while the animal has not moved or minimally). I tried increasing the node sigma a bit but does not seem to help a lot. Perhaps there's something else?

**Comment by talmo** [comment]
Hi @donw7,

Alright, that sounds like progress :)

To reduce the jitter, there's a couple of things you can do.

First is to train with a _lower_ sigma for the confidence maps. This makes it harder to train the confidence map model since there's less density around your points, but in exchange you get more localization precision.

SLEAP does some amount of subpixel refinement by default, but it doesn't help too much when the sigma is too large. The optimal value is ~1.0, but this might make it hard to find tricky body parts, so I'd recommend lowering it to 2.5 or so first.

Second is to try some light smoothing post-hoc. I like to use the [Savitzky-Golay](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.savgol_filter.html) filter:
```py
from scipy.signal import savgol_filter

smoothed_tracks = savgol_filter(pose_tracks, window_length=5, polyorder=9, axis=0)
```

There are more sophisticated approaches, but let's start there and see how we do.

Cheers,

Talmo

**Comment by talmo** [comment]
Hi @donw7,

I'm going to close this issue for now but please feel free to re-open if you'd like to follow up on this discussion some more.

Cheers,

Talmo

**Comment by donw7** [comment]
@talmo thanks so much! I haven’t had a chance to look in detail yet but will keep you posted.
